Src/ST-LIB_LOW/StateMachine/StateMachine.cpp

```diff
@@ -105,6 +109,11 @@ StateMachine::StateMachine(uint8_t initial_state) :
 	initial_state(initial_state), current_state(initial_state) {
 	add_state(initial_state);
 	enter_state(initial_state);
+
+	#ifdef SIM_ON
+		state_machine_id_in_shm = SharedMemory::state_machine_memory[0]++;
+		SharedMemory::update_current_state(state_machine_id_in_shm,initial_state);
+	#endif
```

Comment:
this does not work, lets think of an example:

First execution:
```c++
state_machine_memory[0] = 0 // this was done in shared_memory_start
//now:
// because you are using postfix increment 
// check this example https://godbolt.org/z/Gn6ef4T64
state_machine_id_in_shm = 0; 
// now you call update_current_state
// 0 is your state machine id
// if initial state is 0 for example you have a loop
//if it is different than one, for example, 6 what happens,

state_machine_memory[0] = initial_state

// on the next iteration you have that state_machine_memory[0] = 6
// and so the next id that you will assign will be 6

```